workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectReference.cs

```diff
@@ -2,8 +2,35 @@
 
 namespace Improbable.Gdk.Core
 {
+    /// <summary>
+    ///     Invokes dispatcher callbacks on GameObjects that represent entities.
+    /// </summary>
     public class GameObjectReference : Component
     {
         public GameObject GameObject;
+
+        public void OnAddComponent(uint componentId)
```

Comment:
Will this be a system calling this which then has to call back out to a system to find the relevant data? If so can we cut out the middle man and just call whatever this calls?

Reply:
Re: First question: Yes, that's the current plan. See Option 1 in my comment above.
Re: Second question: That would be Option 3 (at least I don't see any other way to do this without doing Option 3) which I find less preferable. See comment above.

Reply:
Not this would ultimately get rid of the game object reference. The thing I'm questioning is having systems call into Components which call back into systems.

Instead you could just have the original codegen function call all of that.

sorry I didn't see the options as github hides them when the code changes.
Incidentally you can do option 2 by fully qualifying all names. If this is being done in codegen that isn't actually a problem. Although I don't think I'd recommend it as a solution.

Reply:
I'm not sure what your proposed solution is, surely you'd need some component to keep track of what readers/writers could be available and handle injection and activation? That involves storing data about what spatialOS components we already have, what behaviours need what components etc. Unless we read through what we need and what we have and check a massive amount of info whenever components are added/removed...

I suppose after activation the readerwriters themselves could be stored pretty much anywhere, but even there I don't see any simple options that would be a net improvement on the system overall, especially when the only improvement seems to be just one less level of indirection here?

Reply:
Maybe, @samiwh's idea here is have something like this:
```
// codegenerated classes, i.e. GameObjectComponentDispatcher
public override void InvokeOnComponentUpdateCallbacks()
{
    var gameObjectReferences = ComponentsUpdatedComponentGroup.GetComponentArray<GameObjectReference>();
    for (var i = 0; i < gameObjectReferences.Length; i++)
    {
        var readersWriters = gameObjectReferences[i].GetReadersWriters(54 /* the componentId */);
        readersWriters.foreach(readerWriter => readerWriter.OnComponentUpdate(update /* concrete update type */);)
    }
}
```

Reply:
Discussed offline: We decided to go ahead with the store managed data in a system approach.